### PR TITLE
Support doctrine/orm v3 + doctrine/dbal v4

### LINKED ISSUE #1: Support doctrine/orm v3
Release notes: https://github.com/doctrine/orm/releases/tag/3.0.0

`doctrine/orm:3` comes with some BC breaks which need to be considered in this module.

### LINKED ISSUE #4: Deprecation warning: EM->connect public usage
When using Codeception doctrine2 methods, I'm getting this deprecation warning:

>Public access to Connection::connect() is deprecated. (Connection.php:364 called by Doctrine2.php:260, https://github.com/doctrine/dbal/issues/4966, package doctrine/dbal) /var/app/vendor/doctrine/deprecations/lib/Doctrine/Deprecations/Deprecation.php:209

Looks like calling this method directly from `retrieveEntityManager` will be impossible in DBAL 4.0, the method will become protected.

### PR DESCRIPTION
**Pull request by W0rma**
closes Codeception/module-doctrine#1 

closes Codeception/module-doctrine#4 

I cherry-picked the commits from https://github.com/Codeception/module-doctrine2/pull/73 (fixes a deprecation when using ORM v2) and https://github.com/Codeception/module-doctrine2/pull/76 (runs tests with PHP 8.2 and 8.3).

### CODE REVIEW

File: tests/unit/Codeception/Module/Doctrine2Test.php
Comment (W0rma) [inline tests/unit/Codeception/Module/Doctrine2Test.php:578]: [`EntityManager::$repositoryFactory`](https://github.com/doctrine/orm/blob/5b8b5f28f535e1f03b54dcfb0427407ed92f5b72/src/EntityManager.php#L87) was made `readonly` in ORM v3.

File: src/Codeception/Module/Doctrine2.php
Comment (W0rma) [inline src/Codeception/Module/Doctrine2.php]: [`EntityManager::$repositoryFactory`](https://github.com/doctrine/orm/blob/5b8b5f28f535e1f03b54dcfb0427407ed92f5b72/src/EntityManager.php#L87) was made `readonly` in ORM v3.

As a consequence `haveFakeRepository()` does not work for ORM v3 anymore.

**Review by burned42 — APPROVED** [review]

**Comment by W0rma** [comment]
@Naktibalda @TavoNiievez Please let me know if I can help to move this PR forward. It would be great to have support for DBAL v4 and ORM v3.

I have tested the changes in two of my projects and they seem to work fine.
I tested the changes using ORM v2 and v3.

The PRs https://github.com/Codeception/module-doctrine2/pull/73 and https://github.com/Codeception/module-doctrine2/pull/76 can be closed as soon as this PR has been merged.

**Comment by TavoNiievez** [comment]
@W0rma For my part, looking very superficially at the changes.

It seems to me that we should not replace annotations by attributes in all tests. Mapping with annotations is still valid so we should have at least 1 test with them to verify that they still work.

Then, we don't need PHP 8.0 support anymore, since that version is no longer officially supported. So, instead of you creating compatibility logic for PHP 8.0 simply remove its support from composer.json and main.yml.

Finally, I would not want this PR to be released in this repository if it is merged. @ThomasLandauer already talked several times about this and each of those times I agreed with him regarding 'does it make sense to have a module that supports Doctrine 3 but is called module-doctrine2?'. New repositories had to be created for other modules to remove the version number from their name and the same should happen here.

cc. @Naktibalda

**Comment by ThomasLandauer** [comment]
Indeed, since Doctrine ORM 3 is out now, renaming the module from "Doctrine2" to "Doctrine" is more urgent now. Here's the issue you're referring to: https://github.com/Codeception/module-doctrine/issues/27

**Comment by W0rma** [comment]
@TavoNiievez 

> It seems to me that we should not replace annotations by attributes in all tests. Mapping with annotations is still valid so we should have at least 1 test with them to verify that they still work.
> 

Done. Tests now use attributes with ORM v3 and annotations with ORM v2.

> Then, we don't need PHP 8.0 support anymore, since that version is no longer officially supported. So, instead of you creating compatibility logic for PHP 8.0 simply remove its support from composer.json and main.yml.
> 

Done. PHP 8.0 has been dropped.

> Finally, I would not want this PR to be released in this repository if it is merged.

Personally, I don't see a problem with this repository supporting ORM v2 _and_ ORM v3.
Do you now who are the maintainers of this repository?

According to https://github.com/Codeception/Codeception/issues/6733 @Naktibalda does not plan to make bigger contributions anymore. I really hope that there is someone who can care about making this repository compatible with ORM v3.

**Review by TavoNiievez — APPROVED** [review]

**Comment by DavertMik** [comment]
@W0rma do you want to be maintainer of this repository? 😅

**Comment by TavoNiievez** [comment]
A repository has been created that does not include the version number in its name:
https://github.com/Codeception/module-doctrine

I have moved this PR to that repository and already merged it:
https://github.com/Codeception/module-doctrine/pull/26

I will close the PR to avoid duplicity, it only remains to rename the Doctrine2 class to Doctrine and update the references in the documentation to the new repository, when this is done version 3.1.0 of `codeception/module-doctrine` can be released.

**Pull request closed without merging**
